wavespectra/core/watershed.py

```diff
@@ -0,0 +1,209 @@
+        - swells (int): Number of swell partitions to compute.
+        - agefac (float): Age factor.
+        - wscut (float): Wind speed cutoff.
+
+    Returns:
+        - specpart (3darray): Wave spectrum partitions with shape (np, nf, nd).
+
+    """
+    part_array = specpart.partition(spectrum)
+
+    Up = agefac * wspd * np.cos(D2R * (dir - wdir))
+    windbool = np.tile(Up, (freq.size, 1)) > np.tile(
+        celerity(freq, dpt)[:, np.newaxis], (1, dir.size)
+    )
+
+    ipeak = 1  # values from specpart.partition start at 1
+    part_array_max = part_array.max()
+    # partitions_hs_swell = np.zeros(part_array_max + 1)  # zero is used for sea
+    partitions_hs_swell = np.zeros(part_array_max + 1)  # zero is used for sea
+    while ipeak <= part_array_max:
+        part_spec = np.where(part_array == ipeak, spectrum, 0.0)
+
+        # Assign new partition if multiple valleys and satisfying conditions
+        __, imin = inflection(part_spec, freq, dfres=0.01, fmin=0.05)
+        if len(imin) > 0:
+            part_spec_new = part_spec.copy()
+            part_spec_new[imin[0].squeeze() :, :] = 0
+            newpart = part_spec_new > 0
+            if newpart.sum() > 20:
+                part_spec[newpart] = 0
+                part_array_max += 1
+                part_array[newpart] = part_array_max
+                partitions_hs_swell = np.append(partitions_hs_swell, 0)
+
+        # Assign sea partition
+        W = part_spec[windbool].sum() / part_spec.sum()
+        if W > wscut:
+            part_array[part_array == ipeak] = 0
+        else:
+            partitions_hs_swell[ipeak] = hs(part_spec, freq, dir)
+
+        ipeak += 1
+
+    sorted_swells = np.flipud(partitions_hs_swell[1:].argsort() + 1)
+    parts = np.concatenate(([0], sorted_swells[:swells]))
+    all_parts = []
+    for part in parts:
+        all_parts.append(np.where(part_array == part, spectrum, 0.0))
+
+    # Extend partitions list if it is less than swells
+    if len(all_parts) < swells + 1:
+        nullspec = 0 * spectrum
+        nmiss = (swells + 1) - len(all_parts)
+        for i in range(nmiss):
+            all_parts.append(nullspec)
+
+    return np.array(all_parts)
+
+
+def partition(
+    dset,
+    wspd="wspd",
+    wdir="wdir",
+    dpt="dpt",
+    swells=3,
+    agefac=1.7,
+    wscut=0.3333,
+):
+    """Watershed partitioning.
+
+    Args:
+        - dset (xr.DataArray, xr.Dataset): Spectra array or dataset in wavespectra convention.
+        - wspd (xr.DataArray, str): Wind speed DataArray or variable name in dset.
+        - wdir (xr.DataArray, str): Wind direction DataArray or variable name in dset.
+        - dpt (xr.DataArray, str): Depth DataArray or the variable name in dset.
+        - swells (int): Number of swell partitions to compute.
+        - agefac (float): Age factor.
+        - wscut (float): Wind speed cutoff.
+
+    Returns:
+        - dspart (xr.Dataset): Partitioned spectra dataset with extra dimension.
```

Comment:
It looks like the partitions are populated with zero partitions if <swells number of swells are present. Might be worth putting in a note to that effect give the change from max_swells

Reply:
Yep we should made this change more clear. Do you mean put a note in the release notes or in the docstring?

The actual behaviour though has not really changed with the new argument. With the previous version, `max_swells` controlled the maximum number for the full dataset and specific spectrum in the dataset that did not have enough partitions were also filled with zeros. I thought of changing the argument name / description though as we now vectorise the calculations and require a fixed number of partitions so output can be constructed. And `max_swells` is a bit misleading given some spectrum will not have those numbers.